File: src/com/easy/PathSumIII_437.java
/**
 * 
 */
package com.easy;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class PathSumIII_437 {

    // Recursive version
    // Runtime: 11 ms, faster than 64.98%
    // Memory Usage: 42.1 MB, less than 11.25%
    public int pathSum(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        }
        return pathSumStartWithRoot(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
    }

    private int pathSumStartWithRoot(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        }
        int mResult = 0;
        if (root.val == sum) {
            mResult++;
        }
        mResult += pathSumStartWithRoot(root.left, sum - root.val) + pathSumStartWithRoot(root.right, sum - root.val);
        return mResult;
    }

    // DP version
    // Runtime: 3 ms, faster than 100.00%
    // Memory Usage: 40.2 MB, less than 69.00%
    public int pathSum2(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        }
        Map<Integer, Integer> mMap = new HashMap<>();
        mMap.put(0, 1); // initial value
        return pathSumStartWithRoot(root, sum, mMap, 0);
    }

    private int pathSumStartWithRoot(TreeNode root, int sum, Map<Integer, Integer> mMap, int mPreviousDP) {
        if (root == null) {
            return 0;
        }
        int mCurrent = mPreviousDP + root.val;
        int mResult = mMap.getOrDefault(mCurrent - sum, 0);
        mMap.put(mCurrent, mMap.getOrDefault(mCurrent, 0) + 1);
        mResult += pathSumStartWithRoot(root.left, sum, mMap, mCurrent)
                + pathSumStartWithRoot(root.right, sum, mMap, mCurrent);
        mMap.put(mCurrent, mMap.get(mCurrent) - 1); // update to hash map
        return mResult;
    }
}

File: Golang_src/medium/pathSumIII_437.go
package medium

// Use DFS:
// Time Complexity: O(n^2)
// Space Complexity:O(n)
// Runtime: 16 ms, faster than 42.72%
// Memory Usage: 4.4 MB, less than 5.83%
func pathSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}

	count := pathStartSum(root, sum) + pathSum(root.Left, sum) + pathSum(root.Right, sum)

	return count
}

func pathStartSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}
	result := 0

	if root.Val == sum {
		result++
	}

	return result + pathStartSum(root.Left, sum-root.Val) + pathStartSum(root.Right, sum-root.Val)
}

// Use prefix sum:
// Time Complexity: O(n)
// Space Complexity:O(h)
// Runtime: 4 ms, faster than 98.06%
// Memory Usage: 5 MB, less than 5.83%
func pathSum2(root *TreeNode, sum int) int {
	hashMap := map[int]int{0: 1}

	return helper(root, 0, sum, hashMap)
}

func helper(node *TreeNode, curSum, sum int, hashMap map[int]int) int {
	if node == nil {
		return 0
	}
	curSum += node.Val
	result := hashMap[curSum-sum]

	hashMap[curSum]++

	result += helper(node.Left, curSum, sum, hashMap)
	result += helper(node.Right, curSum, sum, hashMap)

	hashMap[curSum]--

	return result
}
